src/hotspot/share/jfr/recorder/stacktrace/jfrAsyncStackTrace.cpp

```diff
@@ -0,0 +1,159 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "jfr/recorder/stacktrace/jfrAsyncStackTrace.hpp"
+#include "jfr/utilities/jfrTypes.hpp"
+#include "jfr/recorder/repository/jfrChunkWriter.hpp"
+#include "jfr/recorder/storage/jfrBuffer.hpp"
+#include "runtime/threadCrashProtection.hpp"
+
+
+
+JfrAsyncStackFrame::JfrAsyncStackFrame(const Method* method, int bci, u1 type, int lineno, const InstanceKlass* ik) :
+  _klass(ik), _method(method), _line(lineno), _bci(bci), _type(type) {}
+
+
+JfrAsyncStackTrace::JfrAsyncStackTrace(JfrAsyncStackFrame* frames, u4 max_frames) :
+  _frames(frames),
+  _nr_of_frames(0),
+  _max_frames(max_frames),
+  _reached_root(false)
+  {}
+
+
+static const size_t min_valid_free_size_bytes = 16;
+
+static inline bool is_full(const JfrBuffer* enqueue_buffer) {
+  return enqueue_buffer->free_size() < min_valid_free_size_bytes;
+}
+
+bool JfrAsyncStackTrace::record_async(JavaThread* jt, const frame& frame) {
+  NoHandleMark nhm;
+
+  assert(jt != nullptr, "invariant");
+  Thread* current_thread = Thread::current_or_null_safe();
+  if (current_thread == nullptr) {
+    return false;
+  }
+  assert(current_thread->in_asgct(), "invariant");
+  assert(jt != current_thread || current_thread->in_asgct(), "invariant");
+
+  u4 count = 0;
+  _reached_root = true;
+
+  JfrVframeStream vfs(jt, frame, false, true);
+
+  while (!vfs.at_end()) {
+    if (count >= _max_frames) {
+      _reached_root = false;
+      break;
+    }
+    const Method* method = vfs.method();
+    if (method == nullptr || !Method::is_valid_method(method)) {
+      // we throw away everything we've gathered in this sample since
+      // none of it is safe
+      return false;
+    }
+    if (!Method::is_valid_method(method)) {
+      // we throw away everything we've gathered in this sample since
+      // none of it is safe
+      return false;
+    }
+    u1 type = vfs.is_interpreted_frame() ? JfrStackFrame::FRAME_INTERPRETER : JfrStackFrame::FRAME_JIT;
+    int bci = 0;
+    if (method->is_native()) {
+      type = JfrStackFrame::FRAME_NATIVE;
+    } else {
+      bci = vfs.bci();
+    }
+
+    intptr_t* frame_id = vfs.frame_id();
+    vfs.next_vframe();
+    if (type == JfrStackFrame::FRAME_JIT && !vfs.at_end() && frame_id == vfs.frame_id()) {
+      // This frame and the caller frame are both the same physical
+      // frame, so this frame is inlined into the caller.
+      type = JfrStackFrame::FRAME_INLINE;
+    }
+    _frames[count] = JfrAsyncStackFrame(method, bci, type, method->line_number_from_bci(bci), method->method_holder());
```

Comment:
Why wouldn't we then do this in parallel in the sampled threads? Especially on systems with lot's of cores, this could otherwise be an issue and drastically increase the amount of dropped samples.

Reply:
Makes sense.